File: packages/template-rust/src/functions/encoding.rs
//! escape, raw, json, url, nl2br, str, type (FUN-10, FUN-11, FUN-18, FUN-19, FUN-26 to FUN-30).

use crate::escape::escape_html;
use crate::functions::helpers::{BuiltIn, FunctionContext, FunctionError, arg_string, stringify_arg};
use crate::value::Value;
use crate::value::number::number_to_string;

fn json_string(text: &str) -> String {
    let mut result = String::from("\"");
    for char in text.chars() {
        match char {
            '"' => result.push_str("\\\""),
            '\\' => result.push_str("\\\\"),
            '\n' => result.push_str("\\n"),
            '\r' => result.push_str("\\r"),
            '\t' => result.push_str("\\t"),
            '\u{8}' => result.push_str("\\b"),
            '\u{c}' => result.push_str("\\f"),
            '<' => result.push_str("\\u003c"),
            '>' => result.push_str("\\u003e"),
            '&' => result.push_str("\\u0026"),
            '\u{2028}' => result.push_str("\\u2028"),
            '\u{2029}' => result.push_str("\\u2029"),
            other if (other as u32) < 0x20 => result.push_str(&format!("\\u{:04x}", other as u32)),
            other => result.push(other),
        }
    }
    result.push('"');
    result
}

/// FUN-26 to FUN-28: compact JSON text of a value.
pub fn to_json(value: &Value) -> String {
    match value {
        Value::Null => "null".to_string(),
        Value::Bool(true) => "true".to_string(),
        Value::Bool(false) => "false".to_string(),
        Value::Number(number) => number_to_string(*number),
        Value::Str(text) | Value::Safe(text) => json_string(text),
        Value::List(list) => format!("[{}]", list.iter().map(to_json).collect::<Vec<_>>().join(",")),
        Value::Map(map) => {
            let parts: Vec<String> = map
                .iter()
                .map(|(key, value)| format!("{}:{}", json_string(key), to_json(value)))
                .collect();
            format!("{{{}}}", parts.join(","))
        }
    }
}

/// FUN-29: percent-encodes every byte except the unreserved characters.
pub fn percent_encode(text: &str) -> String {
    let mut result = String::new();
    for byte in text.bytes() {
        if byte.is_ascii_alphanumeric() || matches!(byte, b'-' | b'_' | b'.' | b'~') {
            result.push(byte as char);
        } else {
            result.push_str(&format!("%{byte:02X}"));
        }
    }
    result
}

fn escape(args: &[Value], _context: &FunctionContext<'_>) -> Result<Value, FunctionError> {
    Ok(Value::safe_text(escape_html(&stringify_arg(&args[0])?).into_owned()))
}

fn raw(args: &[Value], _context: &FunctionContext<'_>) -> Result<Value, FunctionError> {
    Ok(Value::safe_text(stringify_arg(&args[0])?))
}

fn json(args: &[Value], _context: &FunctionContext<'_>) -> Result<Value, FunctionError> {
    Ok(Value::text(to_json(&args[0])))
}

fn url(args: &[Value], _context: &FunctionContext<'_>) -> Result<Value, FunctionError> {
    Ok(Value::text(percent_encode(&stringify_arg(&args[0])?)))
}

fn nl2br(args: &[Value], _context: &FunctionContext<'_>) -> Result<Value, FunctionError> {
    let text = arg_string(&args[0], "nl2br")?;
    Ok(Value::text(text.replace("\r\n", "\n").replace('\n', "<br>\n")))
}

fn str_function(args: &[Value], _context: &FunctionContext<'_>) -> Result<Value, FunctionError> {
    Ok(Value::text(stringify_arg(&args[0])?))
}

fn type_function(args: &[Value], _context: &FunctionContext<'_>) -> Result<Value, FunctionError> {
    Ok(Value::text(args[0].value_type().name().to_string()))
}

/// The encoding function group.
pub fn functions() -> Vec<(&'static str, BuiltIn)> {
    vec![
        (
            "escape",
            BuiltIn {
                min: 1,
                max: 1,
                call: escape,
            },
        ),
        ("raw", BuiltIn { min: 1, max: 1, call: raw }),
        (
            "json",
            BuiltIn {
                min: 1,
                max: 1,
                call: json,
            },
        ),
        ("url", BuiltIn { min: 1, max: 1, call: url }),
        (
            "nl2br",
            BuiltIn {
                min: 1,
                max: 1,
                call: nl2br,
            },
        ),
        (
            "str",
            BuiltIn {
                min: 1,
                max: 1,
                call: str_function,
            },
        ),
        (
            "type",
            BuiltIn {
                min: 1,
                max: 1,
                call: type_function,
            },
        ),
    ]
}

File: packages/template-ts/src/functions/encoding.ts
// escape, raw, json, url, nl2br, str, type (FUN-10, FUN-11, FUN-18, FUN-19, FUN-26 to FUN-30).
import { escapeHtml } from '../escape.js';
import { numberToString } from '../value/number.js';
import { NativeObject, SafeString, typeOf, type Value } from '../value/value.js';
import { argString, safe, stringifyArg, type BuiltIn } from './helpers.js';

function jsonString(text: string): string {
  let result = '"';
  for (const char of text) {
    const code = char.codePointAt(0) as number;
    switch (char) {
      case '"': result += '\\"'; break;
      case '\\': result += '\\\\'; break;
      case '\n': result += '\\n'; break;
      case '\r': result += '\\r'; break;
      case '\t': result += '\\t'; break;
      case '\b': result += '\\b'; break;
      case '\f': result += '\\f'; break;
      case '<': result += '\\u003c'; break;
      case '>': result += '\\u003e'; break;
      case '&': result += '\\u0026'; break;
      case ' ': result += '\\u2028'; break;
      case ' ': result += '\\u2029'; break;
      default:
        if (code < 0x20) result += '\\u' + code.toString(16).padStart(4, '0');
        else result += char;
    }
  }
  return result + '"';
}

export function toJson(value: Value): string {
  if (value === null) return 'null';
  if (typeof value === 'boolean') return value ? 'true' : 'false';
  if (typeof value === 'number') return numberToString(value);
  if (typeof value === 'string') return jsonString(value);
  if (value instanceof SafeString) return jsonString(value.text);
  if (value instanceof NativeObject) return 'null';
  if (Array.isArray(value)) return '[' + value.map(toJson).join(',') + ']';
  const parts: string[] = [];
  for (const [key, entry] of value) parts.push(jsonString(key) + ':' + toJson(entry));
  return '{' + parts.join(',') + '}';
}

export function percentEncode(text: string): string {
  return encodeURIComponent(text).replace(/[!'()*]/g, char => '%' + char.charCodeAt(0).toString(16).toUpperCase());
}

export const encodingFunctions: Record<string, BuiltIn> = {
  escape: { min: 1, max: 1, call: ([v]) => safe(escapeHtml(stringifyArg(v as Value))) },
  raw: { min: 1, max: 1, call: ([v]) => safe(stringifyArg(v as Value)) },
  json: { min: 1, max: 1, call: ([v]) => toJson(v as Value) },
  url: { min: 1, max: 1, call: ([v]) => percentEncode(stringifyArg(v as Value)) },
  nl2br: { min: 1, max: 1, call: ([v]) => argString(v as Value, 'nl2br').replace(/\r\n|\n/g, match => (match === '\n' ? '<br>\n' : '<br>\n')) },
  str: { min: 1, max: 1, call: ([v]) => stringifyArg(v as Value) },
  type: { min: 1, max: 1, call: ([v]) => typeOf(v as Value) },
};
